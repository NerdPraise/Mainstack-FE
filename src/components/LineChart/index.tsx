import { LineChart as RLC, Line, XAxis, ResponsiveContainer } from 'recharts'

const LineChart = ({
  data,
  width,
  height,
}: {
  data: { name: string; uv: number | string; pv: number; amt: number }[]
  width: number | string
  height: number | string
}) => {
  return (
    <ResponsiveContainer width={width} height={height}>
      <RLC data={data}>
        <Line type="monotone" dataKey="pv" stroke="#FF5403" dot={false} />
        <XAxis
          dataKey="uv"
          interval={0}
          axisLine={{ stroke: '#DBDEE5' }}
          tickLine={false}
          tick={(props) => (
            <CustomLineTick {...props} dataLength={data.length} />
          )}
        />
      </RLC>
    </ResponsiveContainer>
  )
}
export default LineChart

const CustomLineTick = ({
  payload,
  index,
  dataLength,
  ...props
}: {
  x: number
  y: number
  payload: { value: string }
  index: number
  dataLength: number
}) => {
  const isFirst = index === 0
  const isLast = index === dataLength - 1

  if (!isFirst && !isLast) return null

  return (
    <text
      {...props}
      y={props.y + 12}
      fill="#5E6C84"
      fontSize={14}
      textAnchor={isFirst ? 'start' : isLast ? 'end' : 'middle'}
    >
      {payload.value}
    </text>
  )
}
