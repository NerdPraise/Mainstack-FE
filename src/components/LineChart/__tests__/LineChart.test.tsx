import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest'
import { render } from '@testing-library/react'
import LineChart from '../index'

const { ResizeObserver } = window

const data = [
  { name: 'Page A', uv: 400, pv: 2400, amt: 2400 },
  { name: 'Page A', uv: 500, pv: 2400, amt: 2400 },
  { name: 'Page A', uv: 100, pv: 2400, amt: 2400 },
  { name: 'Page A', uv: 800, pv: 2400, amt: 2400 },
]

beforeEach(() => {
  delete (window as { ResizeObserver?: typeof ResizeObserver }).ResizeObserver
  window.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
  }))
})

afterEach(() => {
  window.ResizeObserver = ResizeObserver
  vi.restoreAllMocks()
})

describe('LineChart', () => {
  it('renders with default props', () => {
    const { container } = render(
      <LineChart width={500} height={300} data={data} />
    )
    expect(container).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders with percentage dimensions', () => {
    const { container } = render(
      <LineChart width="100%" height="100%" data={data} />
    )
    expect(container).toBeInTheDocument()
  })

  it('renders with numeric dimensions', () => {
    const { container } = render(
      <LineChart width={1000} height={600} data={data} />
    )
    expect(container).toBeInTheDocument()
  })

  it('renders ResponsiveContainer with correct props', () => {
    const { container } = render(
      <LineChart width={500} height={300} data={data} />
    )
    const responsiveContainer = container.querySelector(
      '.recharts-responsive-container'
    )
    expect(responsiveContainer).toBeInTheDocument()
  })

  it('renders LineChart with correct data', () => {
    const { container } = render(
      <LineChart width={500} height={300} data={data} />
    )
    const lineChart = container.querySelector('.recharts-line')
    expect(lineChart).toBeInTheDocument()
  })

  it('renders XAxis with correct props', () => {
    const { container } = render(
      <LineChart width={500} height={300} data={data} />
    )
    const xAxis = container.querySelector('.recharts-xAxis')
    expect(xAxis).toBeInTheDocument()
  })
})
