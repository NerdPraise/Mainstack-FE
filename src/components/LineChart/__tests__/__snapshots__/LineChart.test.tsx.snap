// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LineChart > renders with default props 1`] = `
<div>
  <div
    class="recharts-responsive-container"
    style="width: 500px; height: 300px; min-width: 0;"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 100%; height: 100%; max-height: 300px; max-width: 500px;"
    >
      <svg
        class="recharts-surface"
        height="300"
        style="width: 100%; height: 100%;"
        viewBox="0 0 500 300"
        width="500"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="260"
              width="490"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M5,5C59.44444444444445,5,113.88888888888889,5,168.33333333333334,5C222.7777777777778,5,277.22222222222223,5,331.6666666666667,5C386.11111111111114,5,440.55555555555554,5,495,5"
            fill="none"
            height="260"
            stroke="#FF5403"
            stroke-dasharray="0px 0px"
            stroke-width="1"
            width="490"
          />
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            orientation="bottom"
            stroke="#DBDEE5"
            width="490"
            x="5"
            x1="5"
            x2="495"
            y="265"
            y1="265"
            y2="265"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-xAxis xAxis"
                fill="#5E6C84"
                font-size="14"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="start"
                verticalAnchor="start"
                visibleTicksCount="4"
                width="490"
                x="5"
                y="285"
              >
                400
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            />
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            />
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-xAxis xAxis"
                fill="#5E6C84"
                font-size="14"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="end"
                verticalAnchor="start"
                visibleTicksCount="4"
                width="490"
                x="495"
                y="285"
              >
                800
              </text>
            </g>
          </g>
        </g>
      </svg>
    </div>
  </div>
</div>
`;
